rust/datafusion/src/optimizer/constant_folding.rs

```diff
@@ -0,0 +1,620 @@
+                Operator::Eq => match (&left, &right) {
+                    (
+                        Expr::Literal(ScalarValue::Boolean(l)),
+                        Expr::Literal(ScalarValue::Boolean(r)),
+                    ) => Expr::Literal(ScalarValue::Boolean(Some(
+                        l.unwrap_or(false) == r.unwrap_or(false),
+                    ))),
+                    (Expr::Literal(ScalarValue::Boolean(b)), _)
+                        if right.get_type(schema)? == DataType::Boolean =>
+                    {
+                        match b {
+                            Some(true) => right,
+                            Some(false) | None => Expr::Not(Box::new(right)),
+                        }
+                    }
+                    (_, Expr::Literal(ScalarValue::Boolean(b)))
+                        if left.get_type(schema)? == DataType::Boolean =>
+                    {
+                        match b {
+                            Some(true) => left,
+                            Some(false) | None => Expr::Not(Box::new(left)),
+                        }
+                    }
+                    _ => Expr::BinaryExpr {
+                        left: Box::new(left),
+                        op: Operator::Eq,
+                        right: Box::new(right),
+                    },
+                },
+                Operator::NotEq => match (&left, &right) {
+                    (
+                        Expr::Literal(ScalarValue::Boolean(l)),
+                        Expr::Literal(ScalarValue::Boolean(r)),
+                    ) => Expr::Literal(ScalarValue::Boolean(Some(
+                        l.unwrap_or(false) != r.unwrap_or(false),
+                    ))),
+                    (Expr::Literal(ScalarValue::Boolean(b)), _)
+                        if right.get_type(schema)? == DataType::Boolean =>
+                    {
+                        match b {
+                            Some(false) | None => right,
+                            Some(true) => Expr::Not(Box::new(right)),
+                        }
+                    }
+                    (_, Expr::Literal(ScalarValue::Boolean(b)))
+                        if left.get_type(schema)? == DataType::Boolean =>
+                    {
+                        match b {
+                            Some(false) | None => left,
+                            Some(true) => Expr::Not(Box::new(left)),
+                        }
+                    }
+                    _ => Expr::BinaryExpr {
+                        left: Box::new(left),
+                        op: Operator::NotEq,
+                        right: Box::new(right),
+                    },
+                },
+                _ => Expr::BinaryExpr {
+                    left: Box::new(left),
+                    op: op.clone(),
+                    right: Box::new(right),
+                },
+            }
+        }
+        Expr::Not(expr) => Expr::Not(Box::new(optimize_expr(&expr, schema)?)),
```

Comment:
`not(not(x)) -> x` might be a simple rule to add

Reply:
I have added this, I didn't add it in the first place because it will require a convergence loop to converge, but I guess we can work on that as a follow up PR.